pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,285 @@
+
+	// Validate audit policy semantically
+	auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+	}
+	auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+	if !ok {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+	}
+	errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+	if len(errList) != 0 {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+	}
+
+	// Validate audit policy schema version against k8s version of shoot
+	if isValidVersion, err := IsValidAuditPolicyVersion(shoot.Spec.Kubernetes.Version, schemaVersion); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	} else if !isValidVersion {
+		err := fmt.Errorf("your shoot cluster version %q is not compatible with audit policy version %q", shoot.Spec.Kubernetes.Version, schemaVersion.GroupVersion().String())
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	return acadmission.Allowed("referenced audit policy is valid")
+}
+
+func (h *handler) admitConfigMap(ctx context.Context, request admission.Request) admission.Response {
+	var (
+		oldCm = &corev1.ConfigMap{}
+		cm    = &corev1.ConfigMap{}
+	)
+
+	if err := h.getOldObject(request, oldCm); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if request.Operation != admissionv1.Update {
+		return acadmission.Allowed("operation is not update")
+	}
+
+	if err := h.decoder.Decode(request, cm); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	if !controllerutil.ContainsFinalizer(cm, shoot.FinalizerName) {
+		return acadmission.Allowed("configmap is not referenced by a Shoot")
+	}
+
+	auditPolicy, err := getAuditPolicy(cm)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	oldAuditPolicy, ok := oldCm.Data[auditPolicyConfigMapDataKey]
+	if ok && oldAuditPolicy == auditPolicy {
+		return acadmission.Allowed("audit policy not changed")
+	}
+
+	// Validate audit policy semantically
+	auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+	}
+	auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+	if !ok {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+	}
+	errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+	if len(errList) != 0 {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+	}
```

Comment:
This code is duplicated, right?

Reply:
created a function to increase maintainability eventhough the function is a bit unintuitive in the return values